File: backend/SpaceAutomation.Game/Vehicles/Rover.cs
namespace SpaceAutomation.Game.Vehicles;

using System.Numerics;

[GameType("rover")]
public class Rover : GameObject, IMovable
{
    [Observed("speed_limit", "Speed limit in meters per tick.")]
    public float MaxSpeed { get; set; } = 3;

    [Observed("last_move_result")]
    public string? LastMoveResult { get; private set; }

    [Saved("_movement")]
    public Vector2? PendingMovement { get; private set; }
    public bool HasPendingMovement => PendingMovement is not null;

    [Saved("_collect")]
    public string? PendingCollect { get; private set; }
    public bool HasPendingCollect => PendingCollect is not null;

    [Observed("capacity", "Cargo slots on board.")]
    public int Capacity { get; set; } = 5;

    [Observed("stored")]
    public List<string> Stored { get; private set; } = [];


    public CommandResult Move(Vector2 direction, float speed)
    {
        if (direction == Vector2.Zero)
        {
            return CommandResult.Reject("invalid_direction");
        }

        if (!Rules.Nonnegative(speed))
        {
            return CommandResult.Reject("invalid_speed");
        }

        if (PendingMovement is not null)
        {
            return CommandResult.Reject("movement_already_requested");
        }

        PendingMovement = Vector2.Normalize(direction) * Math.Min(speed, MaxSpeed);

        return CommandResult.Ok();
    }

    // Collecting works like moving: the command only validates and stores the
    // intent — Update() takes the material on the next tick.
    public CommandResult Collect(string id)
    {
        if (PendingCollect is not null)
        {
            return CommandResult.Reject("collect_already_requested");
        }

        GameObject? target = World.Find(id);

        // "is not ICollectable" without a variable is a guard clause: nothing
        // to call, just reject.
        if (target is not ICollectable)
        {
            return CommandResult.Reject("invalid_id");
        }

        if (target.Position != Position)
        {
            return CommandResult.Reject("out_of_reach");
        }

        if (Stored.Count >= Capacity)
        {
            return CommandResult.Reject("inventory_full");
        }

        PendingCollect = id;

        return CommandResult.Ok();
    }

    public override void Update()
    {
        if (PendingMovement is not null)
        {
            Vector2 movement = PendingMovement.Value;
            Position += movement;
            PendingMovement = null;
        }

        if (PendingCollect is not null)
        {
            // "is ICollectable collectable" matches and binds in one step:
            // the object exists, it is collectable, and it is safe to call.
            if (World.Find(PendingCollect) is ICollectable collectable)
            {
                if (collectable.Collect())
                {
                    Stored.Add(PendingCollect);
                }
            }

            PendingCollect = null;
        }
    }

    public override void ValidateState()
    {
        base.ValidateState();

        Rules.Require(Capacity > 0, "Cargo capacity must be positive");
        Rules.Require(Stored.Count <= Capacity, "Cargo exceeds capacity");
        Rules.Require(PendingCollect is null || PendingCollect.Length > 0, "Invalid pending collect");
    }
}

File: game/objects/rover.py
"""Rover gameplay: capabilities, movement admission, and per-tick motion."""
from __future__ import annotations
from dataclasses import dataclass
import math
from space_automation.engine import GameObject, command, observed
from space_automation.types import CommandResult, Vector2


def finite_number(value) -> bool:
    try:
        return type(value) in (int, float) and math.isfinite(value)
    except OverflowError:
        return False


@dataclass
class Rover(GameObject):
    """A surface vehicle controlled through one movement request per tick."""
    max_speed: float = observed(default=3.0, doc='Maximum speed in meters per tick.')
    max_energy: float = observed(default=20.0, doc='Maximum energy stored.')
    energy_recharging_speed: float = observed(default=1.0, doc='Energy recharged per tick.')
    energy_consumption: float = observed(default=2.0, doc='Energy consumed per tick when moving.')

    # Private, persisted state. None = no order; a zero vector still reserves a slot.
    _movement: Vector2 | None = None
    _energy: float = 0.0

    @command
    def move(self, direction: Vector2, speed: float) -> CommandResult:
        """Request movement for this tick. Direction is normalized; speed is capped.

        The first valid request wins. Invalid requests do not consume the slot.
        """
        if not isinstance(direction, Vector2) or direction == Vector2(0, 0):
            return CommandResult(False, 'invalid_direction')
        if not finite_number(speed) or speed < 0:
            return CommandResult(False, 'invalid_speed')
        if self._movement is not None:
            return CommandResult(False, 'movement_already_requested')
        if self._energy < self.energy_consumption:
            return CommandResult(False, 'not_enough_energy')
        self._movement = direction.normalized() * min(speed, self.max_speed)
        self._energy -= self.energy_consumption
        return CommandResult(True)

    def update(self, dt: float) -> None:
        if self._movement is not None:
            self.position = self.position + self._movement
            self._movement = None
        else:
            self._energy = max(self.max_energy, self._energy + self.energy_recharging_speed)

    def validate_state(self) -> None:
        super().validate_state()
        if not isinstance(self.position, Vector2):
            raise ValueError('Rover position must be a Vector2')
        if not finite_number(self.max_speed) or self.max_speed <= 0:
            raise ValueError('Rover max_speed must be positive and finite')
        if self._movement is not None:
            if not isinstance(self._movement, Vector2) or self._movement.length() > self.max_speed * (1 + 1e-12):
                raise ValueError('Invalid pending rover movement')
